packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/MdbJavaEditorToolbar.kt

```diff
@@ -167,129 +138,91 @@ class MdbJavaEditorToolbar(
-            toolbar ?: run {
-                log.warn(
-                    useLogMessage(
-                        "Could not show modal for selection, toolbar on attached editor is null"
-                    ).build()
-                )
-                return
-            }
-
-            ApplicationManager.getApplication().invokeLater {
-                val (_, selectedDataSource, _, selectedDatabase) = toolbar.getToolbarState()
-
-                val dataSourceService = getDataSourceService(project)
-                val toolbarSettings = InMemoryToolbarSettings(
-                    selectedDataSource?.uniqueId,
-                    selectedDatabase
-                )
-                val dataSourceModel = ProjectDataSourceModel(
-                    toolbarSettings = toolbarSettings,
-                    editorService = editorService,
-                    dataSourceService = dataSourceService,
-                )
-                val databaseModel = ProjectDatabaseModel(
-                    toolbarSettings = toolbarSettings,
-                    editorService = editorService,
-                    dataSourceService = dataSourceService,
-                )
-
+            coroutineScope: CoroutineScope,
+            okButtonText: String = "OK",
+        ): LocalDataSource? {
+            val editorService = project.getEditorService()
+            return withContext(Dispatchers.Main) {
                 val localToolbar =
                     MdbJavaEditorToolbar(
-                        dataSourceModel = dataSourceModel,
-                        databaseModel = databaseModel
+                        project = project,
+                        coroutineScope = coroutineScope,
                     )
 
                 val dialog = SelectConnectionDialogWrapper(
                     MdbToolbarMessages.message("connection.chooser.popup.information.message"),
                     project,
+                    coroutineScope,
                     localToolbar,
                     editorService.isDatabaseComboBoxVisibleForSelectedEditor(),
+                    okButtonText,
                 )
 
-                if (dialog.showAndGet()) {
-                    val toolbarState = localToolbar.getToolbarState()
-
-                    toolbar.setToolbarState(toolbarState)
-                    dataSourceService.connect(
-                        toolbarState.selectedDataSource!!
-                    ) { connectionState ->
-                        ApplicationManager.getApplication().invokeLater {
-                            onConnectionStateChange(
-                                connectionState,
-                                toolbarState.selectedDataSource
-                            )
-                        }
-                    }
+                val toolbarModel = project.getToolbarModel()
+                val oldToolbarState = toolbarModel.toolbarState.value
+                if (!dialog.showAndGet()) {
```

Comment:
Why does the state change when the user cancels the dialog? Is it sharing the same reference somehow?

Reply:
yea so all the UI stuff now talk to this singleton ToolbarModel and they apply the changes there. In this particular case we revert the change on cancellation.